pkg/ruler/rule_backup_manager.go

```diff
@@ -0,0 +1,156 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	promRules "github.com/prometheus/prometheus/rules"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
+)
+
+// Implements GroupLoader interface but instead of reading from a file when Load is called, it returns the
+// rulefmt.RuleGroup it has stored
+type loader struct {
+	ruleGroups map[string][]rulefmt.RuleGroup
+}
+
+func (r *loader) Load(identifier string) (*rulefmt.RuleGroups, []error) {
+	return &rulefmt.RuleGroups{
+		Groups: r.ruleGroups[identifier],
+	}, nil
+}
+
+func (r *loader) Parse(query string) (parser.Expr, error) {
+	return parser.ParseExpr(query)
+}
+
+// rulesBackupManager is an in-memory store that holds []promRules.Group of multiple users. It only stores the Groups,
+// it doesn't evaluate them.
+type rulesBackupManager struct {
+	backupRuleGroupsMtx sync.RWMutex
+	backupRuleGroups    map[string][]*promRules.Group
+	cfg                 Config
+
+	logger log.Logger
+
+	backupGroupRulesCount      *prometheus.GaugeVec
+	lastBackupReloadSuccessful *prometheus.GaugeVec
+}
+
+func newRulesBackupManager(cfg Config, logger log.Logger, reg prometheus.Registerer) *rulesBackupManager {
+	return &rulesBackupManager{
+		backupRuleGroups: make(map[string][]*promRules.Group),
+		cfg:              cfg,
+		logger:           logger,
+
+		backupGroupRulesCount: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_rule_group_rules",
+			Help:      "The number of backed up rules",
+		}, []string{"user", "rule_group"}),
+		lastBackupReloadSuccessful: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_last_reload_successful",
+			Help:      "Boolean set to 1 whenever the last configuration reload attempt was successful.",
+		}, []string{"user"}),
+	}
+}
+
+// backUpRuleGroups updates the map[string][]*promRules.Group that the rulesBackupManager stores in memory.
+func (r *rulesBackupManager) backUpRuleGroups(_ context.Context, ruleGroups map[string]rulespb.RuleGroupList) {
+	backupRuleGroups := make(map[string][]*promRules.Group)
+	for user, groups := range ruleGroups {
+		promGroups, err := r.ruleGroupListToPromGroups(user, groups)
+		if err != nil {
+			r.lastBackupReloadSuccessful.WithLabelValues(user).Set(0)
+			level.Error(r.logger).Log("msg", "unable to back up rules", "user", user, "err", err)
+			continue
+		}
+		backupRuleGroups[user] = promGroups
+		r.lastBackupReloadSuccessful.WithLabelValues(user).Set(1)
+	}
+	r.backupRuleGroupsMtx.Lock()
```

Comment:
I think we do need it specially on the part where we update the metrics. We don't want 2 concurrent updates of the metrics to overwrite each other. One has to win.

Reply:
I see, but isn't the backup triggered by `syncRules` and sync rules happens serially? meaning that it will never happen concurrently?

https://github.com/cortexproject/cortex/blob/4b75a5c1a9bf5c9a98299339d86b7b59b63ac5fa/pkg/ruler/ruler.go#L521

Reply:
Updated, removed the lock